MDEV-15564 Avoid table rebuild in ALTER TABLE on collation or charset change

kevgs opened the pull request:
Patch makes it's possible to in some cases 'increase' CHARSET length without
touching any data. Thus, only metadata is changed. And this has no penalty for
future DDL or DML queries. Additionally it optimizes UNIQUE and PRIMARY keys
dropping and creation if charset was changed for one of its key parts but only
if collate change does not require change of order in a key. If collations
are not 'compatible' operation can not be performed instantly.

Created_field::altered: member to easily check that this field was added by user

btr_index_rec_validate(): relax check to allow REDUNDANT field be shorter than
specified in index definition

fix_key_parts(): only fix key parts without checking anything. This is needed early
by INSTANT ALTER CHARSET check

I'm contributing this new code of the whole pull request, including one or several files that are either new files or modified ones, under the BSD-new license.

[Buildbot](http://buildbot.askmonty.org/buildbot/grid?category=main&branch=tt-10.4-MDEV-15564)

kevgs commented on sql/field.h:
```diff
@@ -4877,19 +4877,20 @@ class Create_field :public Column_definition
     change(old_field->field_name),
     field(old_field), option_struct(old_field->option_struct),
-    create_if_not_exists(false)
+    create_if_not_exists(false), altered(false)
```
Added getter and setter.

kevgs commented on mysql-test/suite/innodb/t/instant_alter_charset.test:
```diff
@@ -0,0 +1,411 @@
+alter table rebuild
+  change a a varchar(150) charset latin1 not null default 'asdf',
+  algorithm=inplace;
+--source include/restart_mysqld.inc
```
This was added to check for persistence of data dictionary changes. Now given that test has a lot of combinations speeding it up a little bit with removing a restart seems to be a good idea.

dr-m replied:
I think that it is sufficient to verify the changes by reading from `information_schema.innodb_sys_columns`. Those reflect the persistent InnoDB data dictionary tables without any caching.
Restart and kill+restart would be necessary when modifying the recovery logic, but not for this task.

kevgs commented on mysql-test/suite/innodb/t/instant_alter_charset.test:
```diff
@@ -0,0 +1,411 @@
+--source include/have_innodb.inc
+--source include/innodb_row_format.inc
```
First of all, `innodb_row_format.inc` * `innodb_page_size.inc` * `alter_algorithm.inc` = 60 which is not fun. Also I don't know why it's needed. Could you elaborate on that? `COPY` is used where it's needed for reference f.ex in `latin1_swedish_special_case` tests.

dr-m replied:
OK, maybe `innodb_page_size.inc` is not that useful for this change. `mtr` is primarily a regression test suite, not a stress test. Page sizes should be covered by additional stress testing. They are interesting, because they affect page splits.

I still think we should cover both COPY and INPLACE with `.combinations` and `.rdiff` files, for all 3 `innodb_row_format.combinations`. That would be only 6 combinations.

Because NOCOPY and INSTANT are subsets of INPLACE, I do not see a need to test with them. We would prove with `innodb_instant_alter_column` that every operation was instantaneous unless the COPY algorithm was used.

kevgs commented on storage/innobase/handler/handler0alter.cc:
```diff
@@ -9585,6 +10154,34 @@ commit_cache_rebuild(
+			DBUG_ASSERT(col->len == new_len);
+		}
+	}
+
```
I removed that code. Thanks for noticing.

kevgs commented on storage/innobase/handler/handler0alter.cc:
```diff
@@ -1262,6 +1263,535 @@ check_v_col_in_order(
+	       && instant_drop_add_unique_index_possible_or_irrelevant(
+		   ha_alter_info)
+	       && instant_drop_add_pk_possible_or_irrelevant(ha_alter_info);
+}
```
Moved to `field.h`, `sql_table.h`.

kevgs commented on sql/field.h:
```diff
@@ -4901,20 +4901,35 @@ class Create_field :public Column_definition
+  {
+    DBUG_ASSERT(field);
+    this->altered= altered;
+  }
```
I've removed `Create_field::altered` completely and use `Field::is_equal()` instead. Thanks for noticing.

dr-m commented on sql/sql_type.cc:
```diff
@@ -7845,3 +7845,84 @@ Type_handler_time_common::create_literal_item(THD *thd,
+        (!strcmp(old_csname, "ascii") && !strcmp(new_csname, "koi8u"))))
+  {
+    return false;
+  }
```
Can you invoke `strcmp(old_csname, "ascii")` only once, and write a bunch of `(!strcmp(new_csname, …) || !strcmp(new_csname, …))`?

Are these really all the character sets that are proper supersets of `"ascii"`? Are `"latin2"` to `"latin7"` not proper supersets of ASCII? Also GBK, UJIS, SJIS, BIG5 would seem to be supersets of ASCII.

Do we support converting anything to `"binary"` (`my_charset_bin->number == 63`) without any validation? If we do, then we should also support that one instantly.

Please confirm your findings with @abarkov.

kevgs replied:
> Can you invoke strcmp(old_csname, "ascii") only once, and write a bunch of (!strcmp(new_csname, …) || !strcmp(new_csname, …))?

> Are these really all the character sets that are proper supersets of "ascii"? Are "latin2" to "latin7" not proper supersets of ASCII? Also GBK, UJIS, SJIS, BIG5 would seem to be supersets of ASCII.

This is fixed and added.

> Do we support converting anything to "binary" (my_charset_bin->number == 63) without any validation? If we do, then we should also support that one instantly.

No, we don't support it.

vaintroub replied:
As a matter of fact, all character sets, with exception of really exotic SWE7, EBCDIC variants, and UCS/UTF<more-than-8>,  are superset of ASCII. pinging  @abarkov , who may want to correct this statement

kevgs replied:
> Do we support converting anything to `"binary"` (`my_charset_bin->number == 63`) without any validation? If we do, then we should also support that one instantly.

Nope. It's disallowed by server code.

abarkov replied:
Conversion to "binary" should be supported without validation.

Do you have an example script that demonstrates it's not supported?
You might hit some bug.

kevgs replied:
Here is the test for that case https://github.com/MariaDB/server/blob/10.4/mysql-test/suite/innodb/t/instant_alter_charset.test#L143

dr-m commented on sql/sql_table.h:
```diff
@@ -285,4 +286,61 @@ extern mysql_mutex_t LOCK_gdl;
+@result	true if condition succedded and false otherwise */
+bool instant_alter_charset_possible_or_irrelevant(
+    const Alter_inplace_info *ha_alter_info);
+
```
I think that these predicates would better be declared as member functions of `class Alter_inplace_info` in `handler.h`.

kevgs commented:
I updated the PR. Moved code to use `IS_EQUAL_PACK_LENGTH`.

Also, current version seems to implement MDEV-16849. Relevant code is in `fill_alter_inplace_info()`.

Still need to move logic to `innobase_instant_try()`.

dr-m commented on storage/innobase/handler/handler0alter.cc:
```diff
@@ -10161,6 +10285,13 @@ commit_try_norebuild(
+	}
+
 	DBUG_RETURN(innobase_instant_try(ha_alter_info, ctx, altered_table,
 					 old_table, trx));
```
[Like I wrote on November 9](https://github.com/MariaDB/server/pull/921/files#r232204607), I’d rather construct `ctx->instant_table` with all the changes and then let `innobase_instant_try()` update `SYS_COLUMNS` and other tables. We should also follow this approach for renaming and extending columns, and remove existing functions like `innobase_rename_or_enlarge_columns_cache()`.

dr-m commented on mysql-test/suite/innodb/t/instant_alter_charset.test:
```diff
@@ -0,0 +1,527 @@
+select * from compressed;
+check table compresed;
+
+drop table compressed;
```
Could we preserve all tables where we did an `ALGORITHM=INSTANT` change until the end of the test?
At the end of the test:
1. Show that no `id` changed for the table or tablespace.
1. Restart the server.
1. Read all data from each table.
1. Drop the tables.

dr-m commented on mysql-test/suite/innodb/t/instant_alter_charset.test:
```diff
@@ -0,0 +1,527 @@
+drop table regression;
+
+# ROW_FORMAT=DYNAMIC limitation:
+# size in bytes cannot be increased from less of equal that 255 to more than 255
```
This is a too strict limitation.
We can allow the maximum size to be increased from less than 128 bytes to anything (even beyond 256 bytes).

dr-m commented on mysql-test/suite/innodb/t/instant_alter_charset.test:
```diff
@@ -0,0 +1,527 @@
+);
+
+insert into fully_compatible (from_charset, from_collate, to_charset, to_collate) values
+  ('utf8mb3', 'utf8mb3_general_ci',           'utf8mb4', 'utf8mb4_general_ci'),
```
Could we declare `from_charset` and `to_charset` virtual columns, so that the `INSERT` statement can only populate the collations?
Can we use `CREATE TEMPORARY TABLE` for this metadata?

dr-m commented on mysql-test/suite/innodb/t/instant_alter_charset.test:
```diff
@@ -0,0 +1,527 @@
+  ('ascii',   'ascii_general_ci',      'latin2',  'latin2_croatian_ci'),
+  ('ascii',   'ascii_general_ci',      'latin7',  'latin7_estonian_cs'),
+  ('utf16',   'utf16_general_ci',      'utf16',   'utf16_german2_ci')
+;
```
These do not seem to be all the combinations. A comment about this would be good.

dr-m commented:
I rebased this, refactored the InnoDB code (mostly due to conflicts with MDEV-15563) and pushed to 10.4 as commit ea0be9e2d63467064292559762887495772240fc.
